### PR TITLE
Contracts & Harnesses for `byte_add`, `byte_offset`, and `byte_offset_from`

### LINKED ISSUE #53: Challenge 6: Safety of `NonNull`
[Challenge 6: Safety of `NonNull`](https://model-checking.github.io/verify-rust-std/challenges/0006-nonnull.html#challenge-6-safety-of-nonnull)

### PR DESCRIPTION
**Pull request by danielhumanmod**
# Description:

This PR introduces function contracts and proof harness for the NonNull pointer in the Rust core library. Specifically, it verifies three new APIs—`byte_offset`, `byte_add`, and `byte_offset_from` with Kani. These changes enhance the functionality of pointer arithmetic and verification for NonNull pointers.

# Changes Overview:

**Covered APIs:**
1. `NonNull::byte_add`: Adds a specified byte offset to a pointer.
2. `NonNull::byte_offset`: Allows calculating an offset from a pointer in bytes.
4. `NonNull::byte_offset_from`: Calculates the distance between two pointers in bytes.

**Proof harness:**
1. `non_null_byte_add_proof`
2. `non_null_byte_offset_proof`
3. `non_null_byte_offset_from_proof`

Towards #53 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: library/core/src/ptr/non_null.rs [hunk 1]
```diff
@@ -502,6 +502,11 @@ impl<T: ?Sized> NonNull<T> {
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_const_stable(feature = "non_null_convenience", since = "1.80.0")]
+    #[requires(!origin.as_ptr().is_null())]
```
Comment (QinyuanWu) [inline library/core/src/ptr/non_null.rs:505]: `!self.as_ptr().is_null()` is not necessary because NonNull guarentees no null pointer. Please consider other properties.

File: library/core/src/ptr/non_null.rs [hunk 2]
```diff
@@ -502,6 +502,11 @@ impl<T: ?Sized> NonNull<T> {
...
+    #[ensures(
```
Comment (QinyuanWu) [inline library/core/src/ptr/non_null.rs:506]: Consider checking that the two pointers are still pointing to the same object(when this approach becomes available)

File: library/core/src/ptr/non_null.rs [hunk 3]
```diff
@@ -579,6 +584,11 @@ impl<T: ?Sized> NonNull<T> {
     #[rustc_allow_const_fn_unstable(set_ptr_value)]
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_const_stable(feature = "non_null_convenience", since = "1.80.0")]
+    #[requires(!self.as_ptr().is_null())]
```
Comment (QinyuanWu) [inline library/core/src/ptr/non_null.rs:587]: same comments as `bytes_offset`

File: library/core/src/ptr/non_null.rs [hunk 4]
```diff
@@ -783,6 +793,11 @@ impl<T: ?Sized> NonNull<T> {
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_const_stable(feature = "non_null_convenience", since = "1.80.0")]
+    #[requires(!origin.as_ptr().is_null())]
```
Comment (QinyuanWu) [inline library/core/src/ptr/non_null.rs:796]: same comments as `bytes_offset`, need to add precondition that ensures the two pointers are pointing to the same object and within bound

File: library/core/src/ptr/non_null.rs [hunk 5]
```diff
@@ -1803,4 +1818,63 @@ mod verify {
...
+        let arr: [i32; ARR_SIZE] = kani::any();
+        let offset = kani::any_where(|x| *x <= ARR_SIZE);
+        let raw_ptr: *mut i32 = arr.as_ptr() as *mut i32;
+        let ptr = unsafe { NonNull::new(raw_ptr.add(offset)).unwrap() };
+        let count: usize = kani::any();
+
+        kani::assume(count < usize::MAX);
+        kani::assume(count.checked_mul(mem::size_of::<i32>()).is_some());
+        kani::assume(count * mem::size_of::<i32>() <= (isize::MAX as usize));
+        kani::assume(count < ARR_SIZE - offset);
```
Comment (QinyuanWu) [inline library/core/src/ptr/non_null.rs:1834]: These assumption should ideally be expressed in the contracts. Could you raise this issue of not being able to use `mem::size_of` in the contracts in the GitHub discussion?

File: library/core/src/ptr/non_null.rs [hunk 6]
```diff
@@ -1803,4 +1818,63 @@ mod verify {
...
+        let offset = kani::any_where(|x| *x <= ARR_SIZE);
+        let raw_ptr: *mut i32 = arr.as_ptr() as *mut i32;
+        let ptr = unsafe { NonNull::new(raw_ptr.add(offset)).unwrap() };
+        let count: isize = kani::any();
+
+        kani::assume(count >= isize::MIN);
+        kani::assume(count <= isize::MAX);
+        kani::assume(count.checked_mul(mem::size_of::<i32>() as isize).is_some());
+        kani::assume(count * (mem::size_of::<i32>() as isize) <= (isize::MAX as isize));
+        kani::assume((offset as isize + count) < (ARR_SIZE as isize));
```
Comment (QinyuanWu) [inline library/core/src/ptr/non_null.rs:1854]: same comment as `byte_add` harness

File: library/core/src/ptr/non_null.rs [hunk 7]
```diff
@@ -1803,4 +1818,63 @@ mod verify {
...
+    }
+
+    #[kani::proof_for_contract(NonNull::byte_offset_from)]
+    pub fn non_null_byte_offset_from_proof() {
+        const ARR_SIZE: usize = 100000;
+        let arr: [i32; ARR_SIZE] = kani::any();
+
+        // Randomly generate offsets for the pointers
+        let offset = kani::any_where(|x| *x < ARR_SIZE);
+        let origin_offset = kani::any_where(|x| *x < ARR_SIZE);
```
Comment (QinyuanWu) [inline library/core/src/ptr/non_null.rs:1868]: This eliminates the situation where one of the pointers is outside of bound of memory or pointing to a different object. This should be eliminated by the contract instead of in the harness.

**Review by QinyuanWu — CHANGES_REQUESTED** [review]

File: library/core/src/ptr/non_null.rs (hunk 1 above)
Comment (danielhumanmod) [reply library/core/src/ptr/non_null.rs:505]: After discuss with sponsor on meeting, we should keep it.

File: library/core/src/ptr/non_null.rs (hunk 2 above)
Comment (danielhumanmod) [reply library/core/src/ptr/non_null.rs:506]: TODO: move it in contract after using latest kani built from source code

File: library/core/src/ptr/non_null.rs [hunk 8]
```diff
@@ -1803,4 +1824,63 @@ mod verify {
...
+        kani::assume((offset as isize + count) < (ARR_SIZE as isize));
+
+        unsafe {
+            let result = ptr.byte_offset(count);
+        }
+    }
+
+    #[kani::proof_for_contract(NonNull::byte_offset_from)]
+    pub fn non_null_byte_offset_from_proof() {
+        const ARR_SIZE: usize = 100000;
```
Comment (zhassan-aws) [inline library/core/src/ptr/non_null.rs:1869]: The harness constrains the two pointers to be from the same allocated object which is undesirable. The `PointerGenerator` can be used to remove this constraint and instead rely on a `requires` clause to enforce this restriction. See https://github.com/model-checking/kani/blob/main/library/kani_core/src/arbitrary/pointer.rs#L142 for an example.

**Comment by danielhumanmod** [comment]
Thanks for your suggestion @zhassan-aws! I've updated the code in the latest commit based on your comments. I’d appreciate it if you could take a look when you have a chance.

File: library/core/src/ptr/non_null.rs [hunk 9]
```diff
@@ -502,6 +502,8 @@ impl<T: ?Sized> NonNull<T> {
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_const_stable(feature = "non_null_convenience", since = "1.80.0")]
+    #[kani::requires(kani::mem::same_allocation(self.as_ptr() as *const(), self.as_ptr().byte_offset(count) as *const()))]
```
Comment (zhassan-aws) [inline library/core/src/ptr/non_null.rs:505]: Is the cast to `*const()` necessary?

File: library/core/src/ptr/non_null.rs [hunk 10]
```diff
@@ -1803,4 +1819,67 @@ mod verify {
...
+    #[kani::proof_for_contract(NonNull::byte_add)]
+    pub fn non_null_byte_add_proof() {
+        const ARR_SIZE: usize = 100000;
+        let arr: [i32; ARR_SIZE] = kani::any();
+        let offset = kani::any_where(|x| *x <= ARR_SIZE);
+        let raw_ptr: *mut i32 = arr.as_ptr() as *mut i32;
+        let ptr = unsafe { NonNull::new(raw_ptr.add(offset)).unwrap() };
+        let count: usize = kani::any();
+
+        kani::assume(count < usize::MAX);
```
Comment (zhassan-aws) [inline library/core/src/ptr/non_null.rs:1832]: What is the purpose of this assumption?

File: library/core/src/ptr/non_null.rs [hunk 11]
```diff
@@ -1803,4 +1819,67 @@ mod verify {
...
+    #[kani::proof_for_contract(NonNull::byte_offset)]
+    pub fn non_null_byte_offset_proof() {
+        const ARR_SIZE: usize = 100000;
+        let arr: [i32; ARR_SIZE] = kani::any();
+        let offset = kani::any_where(|x| *x <= ARR_SIZE);
+        let raw_ptr: *mut i32 = arr.as_ptr() as *mut i32;
+        let ptr = unsafe { NonNull::new(raw_ptr.add(offset)).unwrap() };
+        let count: isize = kani::any();
+
+        kani::assume(count >= isize::MIN);
```
Comment (zhassan-aws) [inline library/core/src/ptr/non_null.rs:1850]: This assumption and the next one are not needed as `kani::any()` when called for `isize` would only produce valid `isize` values.

File: library/core/src/ptr/non_null.rs [hunk 12]
```diff
@@ -1803,4 +1819,67 @@ mod verify {
...
+        unsafe {
+            let result = ptr.byte_offset(count);
+        }
+    }
+
+    #[kani::proof_for_contract(NonNull::byte_offset_from)]
+    pub fn non_null_byte_offset_from_proof() {
+        use kani::PointerGenerator;
+        const SIZE: usize = mem::size_of::<i32>();
+        let mut generator1 = PointerGenerator::<SIZE>::new();
```
Comment (zhassan-aws) [inline library/core/src/ptr/non_null.rs:1863]: We should use a bigger value, e.g. `SIZE * 10` so that the generator can produce distinct pointers.

File: library/core/src/ptr/non_null.rs (hunk 9 above)
Comment (danielhumanmod) [reply library/core/src/ptr/non_null.rs:505]: > Is the cast to `*const()` necessary?

Good point! The reason I did the const cast here is because previously `same_allocation()` is not support `Size?` input, and I got a workaround suggestion from https://github.com/model-checking/verify-rust-std/discussions/141

File: library/core/src/ptr/non_null.rs [hunk 13]
```diff
@@ -579,6 +581,14 @@ impl<T: ?Sized> NonNull<T> {
     #[rustc_allow_const_fn_unstable(set_ptr_value)]
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_const_stable(feature = "non_null_convenience", since = "1.80.0")]
+    #[kani::requires(kani::mem::same_allocation(
+        self.as_ptr() as *const(),
+        ((self.as_ptr() as *const () as usize) + count) as *const()
```
Comment (zhassan-aws) [inline library/core/src/ptr/non_null.rs:586]: Can we use `.addr()` instead of casting to `usize`?

Comment (zhassan-aws) [inline library/core/src/ptr/non_null.rs:586]: Also, I think there's a possibility of overflow in the addition (depending on the pointer value and the `count`). So, we should use `checked_add` to make sure the operation doesn't overflow.

File: library/core/src/ptr/non_null.rs [hunk 14]
```diff
@@ -1803,4 +1819,64 @@ mod verify {
...
+    #[kani::proof_for_contract(NonNull::byte_add)]
+    pub fn non_null_byte_add_proof() {
+        const ARR_SIZE: usize = 100000;
+        let arr: [i32; ARR_SIZE] = kani::any();
+        let offset = kani::any_where(|x| *x <= ARR_SIZE);
+        let raw_ptr: *mut i32 = arr.as_ptr() as *mut i32;
+        let ptr = unsafe { NonNull::new(raw_ptr.add(offset)).unwrap() };
+        let count: usize = kani::any();
+
+        kani::assume(count.checked_mul(mem::size_of::<i32>()).is_some());
```
Comment (zhassan-aws) [inline library/core/src/ptr/non_null.rs:1832]: What is the purpose of these assumptions? The `count` that `byte_add` takes is in bytes, so no multiplication by `mem::size_of::<T>` occurs.

File: library/core/src/ptr/non_null.rs [hunk 15]
```diff
@@ -1803,4 +1819,64 @@ mod verify {
...
+        kani::assume(count.checked_mul(mem::size_of::<i32>() as isize).is_some());
+        kani::assume(count * (mem::size_of::<i32>() as isize) <= (isize::MAX as isize));
+        unsafe {
+            let result = ptr.byte_offset(count);
+        }
+    }
+
+    #[kani::proof_for_contract(NonNull::byte_offset_from)]
+    pub fn non_null_byte_offset_from_proof() {
+        use kani::PointerGenerator;
```
Comment (QinyuanWu) [inline library/core/src/ptr/non_null.rs:1858]: Since other harnesses may also use the PointerGenerator, can we move this use statement to the top of the verification module(line 1804)?

File: library/core/src/ptr/non_null.rs [hunk 16]
```diff
@@ -502,6 +502,13 @@ impl<T: ?Sized> NonNull<T> {
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_const_stable(feature = "non_null_convenience", since = "1.80.0")]
+    #[kani::requires(
+        count <= isize::MAX &&
```
Comment (zhassan-aws) [inline library/core/src/ptr/non_null.rs:506]: This conjunct is a no-op since `count` is an `isize` so it cannot exceed `isize::MAX`.

File: library/core/src/ptr/non_null.rs [hunk 17]
```diff
@@ -502,6 +502,13 @@ impl<T: ?Sized> NonNull<T> {
...
+        count <= isize::MIN &&
```
Comment (zhassan-aws) [inline library/core/src/ptr/non_null.rs:507]: There should be no need to put any restrictions on `count` other than the ones below.

**Review by zhassan-aws — APPROVED** [review]
Nice. Thanks!

File: library/core/src/ptr/non_null.rs [hunk 18]
```diff
@@ -1803,4 +1823,58 @@ mod verify {
...
+    pub fn non_null_byte_offset_from_proof() {
+        const SIZE: usize = mem::size_of::<i32>() * 10;
+        let mut generator1 = PointerGenerator::<SIZE>::new();
+        let mut generator2 = PointerGenerator::<SIZE>::new();
+
+        let ptr: *mut i32 = if kani::any() {
+            generator1.any_in_bounds().ptr as *mut i32
+        } else {
+            generator2.any_in_bounds().ptr as *mut i32
+        };
```
Comment (celinval) [inline library/core/src/ptr/non_null.rs:1865]: FYI, I don't think need this non-determinism here.

File: library/core/src/ptr/non_null.rs [hunk 19]
```diff
@@ -1803,4 +1823,58 @@ mod verify {
         let maybe_null_ptr =  if kani::any() { xptr as *mut i32 } else { null_mut() };
         let _ = NonNull::new(maybe_null_ptr);
     }
+
+    #[kani::proof_for_contract(NonNull::byte_add)]
+    pub fn non_null_byte_add_proof() {
+        const ARR_SIZE: usize = 100000;
+        let arr: [i32; ARR_SIZE] = kani::any();
+        let offset = kani::any_where(|x| *x <= ARR_SIZE);
```
Comment (celinval) [inline library/core/src/ptr/non_null.rs:1831]: I would prefer using the pointer generator everywhere if possible since it will also generate unaligned pointers.

**Review by celinval — COMMENTED** [review]
Please replace `kani::requires` and `kani::ensures` by `safety::requires` and `safety::ensures`. Thanks!

File: library/core/src/ptr/non_null.rs [hunk 20]
```diff
@@ -587,7 +585,6 @@ impl<T: ?Sized> NonNull<T> {
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_const_stable(feature = "non_null_convenience", since = "1.80.0")]
     #[kani::requires(
-        count <= (isize::MAX as usize) &&
```
Comment (zhassan-aws) [inline library/core/src/ptr/non_null.rs]: Why was this removed? I think this is correct since `count` is a `usize`.

Comment (danielhumanmod) [reply library/core/src/ptr/non_null.rs]: Sorry, I delete this line by mistake. Already add it back in latest commit

File: library/core/src/ptr/non_null.rs [hunk 21]
```diff
@@ -579,6 +584,14 @@ impl<T: ?Sized> NonNull<T> {
     #[rustc_allow_const_fn_unstable(set_ptr_value)]
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_const_stable(feature = "non_null_convenience", since = "1.80.0")]
+    #[kani::requires(
+        self.as_ptr().addr().checked_add(count).is_some() &&
+        kani::mem::same_allocation(self.as_ptr() as *const(),((self.as_ptr().addr()) + count) as *const())
+    )]
+    #[kani::ensures(
+        |result: &NonNull<T>|
+        result.as_ptr().addr() == (self.as_ptr().addr() + count)
```
[the remark this answers is not in the document]
Comment (danielhumanmod) [reply library/core/src/ptr/non_null.rs:594]: > Can you please be consistent with your previous contract?

Thanks for the review and suggestions @celinval ! I have solved most of your comments, but not sure I understand this one. Do you mind clarifying it?

File: library/core/src/ptr/non_null.rs (hunk 18 above)
Comment (celinval) [reply library/core/src/ptr/non_null.rs:1866]: Yes, but I think just one of the pointers need that, right? Not a big deal though

Comment (zhassan-aws) [reply library/core/src/ptr/non_null.rs:1866]: Oh, I see what you mean. Makes sense.

File: library/core/src/ptr/non_null.rs [hunk 22]
```diff
@@ -502,6 +502,11 @@ impl<T: ?Sized> NonNull<T> {
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_const_stable(feature = "non_null_convenience", since = "1.80.0")]
+    #[requires(
+        (self.as_ptr().addr() as isize).checked_add(count).is_some() &&
+        kani::mem::same_allocation(self.as_ptr() as *const(), self.as_ptr().byte_offset(count) as *const())
```
Comment (celinval) [inline library/core/src/ptr/non_null.rs:507]: This line is still using `byte_offset`. Since this is a pre-condition, it will be UB if you go out of bounds.

File: library/core/src/ptr/non_null.rs [hunk 23]
```diff
@@ -529,6 +529,11 @@ impl<T: ?Sized> NonNull<T> {
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_const_stable(feature = "non_null_convenience", since = "1.80.0")]
+    #[requires(
```
Comment (celinval) [inline library/core/src/ptr/non_null.rs:532]: I believe it's safe to call this function with `count == 0` even if `self.pointer` is dangling.

File: library/core/src/ptr/non_null.rs [hunk 24]
```diff
@@ -610,6 +615,10 @@ impl<T: ?Sized> NonNull<T> {
...
+        count <= (isize::MAX as usize) - self.as_ptr().addr() &&  // Ensure the offset doesn't overflow
+        kani::mem::same_allocation(self.as_ptr(),self.as_ptr().wrapping_byte_add(count)) // Ensure the offset is within the same allocation
```
Comment (celinval) [inline library/core/src/ptr/non_null.rs:620]: Please use the `ub_checks::same_allocation` instead of `kani`

File: library/core/src/ptr/non_null.rs [hunk 25]
```diff
@@ -812,6 +821,11 @@ impl<T: ?Sized> NonNull<T> {
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_const_stable(feature = "non_null_convenience", since = "1.80.0")]
+    #[requires(kani::mem::same_allocation(self.as_ptr() as *const(), origin.as_ptr() as *const()))]
```
Comment (celinval) [inline library/core/src/ptr/non_null.rs:824]: This should also be safe if the pointers hold the same address, even if they are dangling

**Review by celinval — COMMENTED** [review]
Looks good. Can you please add harnesses for the dangling pointers cases?

File: library/core/src/ptr/non_null.rs (hunk 23 above)
Comment (danielhumanmod) [reply library/core/src/ptr/non_null.rs:532]: > I believe it's safe to call this function with `count == 0` even if `self.pointer` is dangling.

Appreciate the suggestion @celinval! But there is some failure with the my dangling proof:

Proof harness:
```
#[kani::proof_for_contract(NonNull::byte_add)]
    pub fn non_null_byte_add_dangling_proof() {
        let ptr = NonNull::<i32>::dangling();
        let count: usize = kani::any();
        unsafe {
            let _ = ptr.byte_add(count);
        }
    }
```

Error:
```
VERIFICATION RESULT:
 ** 1 of 178 failed
Failed Checks: Kani does not support reasoning about pointer to unallocated memory
 File: "/Users/danieltu/Developer/verify-rust-std/library/core/src/lib.rs", line 423, in kani::mem::cbmc::same_allocation::<i32>

VERIFICATION:- FAILED
Verification Time: 0.26384938s
```

I am not sure whether it is because I didn't create a proper dangling scenario, could you provide some suggestion on this? Thanks!

Comment (celinval) [reply library/core/src/ptr/non_null.rs:532]: Did you extend the requires clause?

File: library/core/src/ptr/non_null.rs (hunk 25 above)
Comment (celinval) [reply library/core/src/ptr/non_null.rs:824]: Please see my question above.

File: library/core/src/ptr/non_null.rs (hunk 23 above)
Comment (danielhumanmod) [reply library/core/src/ptr/non_null.rs:532]: > Did you extend the requires clause?

Hi @celinval, do you mind being more specific about it? Based on my understanding, we need to add a proof harness where `ptr` is dangling and `count` = 0. Any additional contract need to be added into the API?

**Comment by feliperodri** [comment]
@danielhumanmod could you resolve the conflicts?

**Comment by danielhumanmod** [comment]
> @danielhumanmod could you resolve the conflicts?

Hi @feliperodri , I have solved the conflicts in latest commit, thanks for reminding that

File: library/core/src/ptr/non_null.rs (hunk 23 above)
Comment (zhassan-aws) [reply library/core/src/ptr/non_null.rs:535]: I think what @celinval meant is to change:
```rust
kani::mem::same_allocation(self.as_ptr(), self.as_ptr().wrapping_byte_offset(count))
```
to:
```rust
(count == 0 || kani::mem::same_allocation(self.as_ptr(), self.as_ptr().wrapping_byte_offset(count)))
```
so that `same_allocation` is not invoked if `count` is zero.

**Comment by danielhumanmod** [comment]
Thanks for the clarification @zhassan-aws ! I have updated the code based on the comments, appreciate any further or review that might need @zhassan-aws @celinval

**Review by zhassan-aws — APPROVED** [review]
Thanks!

File: library/core/src/ptr/non_null.rs [hunk 26]
```diff
@@ -2300,4 +2318,71 @@ mod verify {
             let distance = ptr_nonnull.offset_from(origin_nonnull);
         }
     }
+
+    #[kani::proof_for_contract(NonNull::byte_add)]
+    pub fn non_null_byte_add_proof() {
+        const ARR_SIZE: usize = mem::size_of::<i32>() * 1000;
```
Comment (tautschnig) [inline library/core/src/ptr/non_null.rs:2324]: Could you please add a comment explaining why you deem "1000" to be the right number?

Comment (danielhumanmod) [reply library/core/src/ptr/non_null.rs:2326]: > Could you please add a comment explaining why you deem "1000" to be the right number?

Thanks for the suggestion @tautschnig, I have updated the code in latest commit

**Review by tautschnig — APPROVED** [review]
